### DISCUSSION TITLE
关于内存中的玩家数据写回mysql

### DISCUSSION THREAD

**Discussion by ericzhanchina** [cloudwu/skynet#1784] [General]
大家谈论的方案，多是玩家登录的时候将数据写入到内存，后续都是操作内存的数据，定期写回到mysql等数据库，那么如何处理服务器断电的情况呢，如果断电玩家的数据可能就会丢失了？

**Comment by zengzizhao** [comment]
如果这样的事件也考虑，是不是还要担心服务器硬盘坏了、服务器CPU烧了、服务器被水淹了之类的，可以不考虑这些问题，因为真出现了你也没任何办法。如果服务器断电的事情都发生了，那就数据回档啊。

**Reply by ericzhanchina** [reply]
数据回档的话，玩家的数据就丢失了，如果服务器每5分钟回写一下数据，那差不多每个玩家都会丢掉5分钟的数据。

兄弟你的服务器都是直接读取到内存，定期写回没有任何的保护吗？

这个断电时间还真是可能发生的，即便买的阿里云的ECS，我都有服务器半夜不知道为何就重启了。

**Comment by zengzizhao** [comment]
我们项目组几个游戏20多台阿里云的服务器，这几年没遇到过服务器断电的情况。
你如果担心断电，那只有即时写回数据库，不过即便如此，因为你也无法避免突然停电造成的那瞬间的数据丢失。

**Reply by ericzhanchina** [reply]
立刻写回恐怕不行，一般数据库也扛不住这样的写操作，虽然断电概率很低，但是确实遇到了，不知道有没有其他的兄弟考虑过这种情况，是用什么方案来保证响应的性能，又可以在服务器突发意外还能保证玩家的数据不丢失。

**Reply by sniper00** [reply]
只能减缓，不能完全避免，除非你的逻辑都用储存过程写。可以先把数据写磁盘上，然后慢慢存db ，这样可以做到毫秒级丢失，并且数据库压力不太大。

**Reply by ericzhanchina** [reply]
大佬都说不太好处理，那看来不容易处理了，尽量减少可能的损失还是很必要的。

**Reply by terry8210** [reply]
看配置能知道DB的测试出极限TPS，然后取一个中位数，分批分量落地，不能保证100%的数据安全，如果担心数据丢失，可以考虑共享内存的方式，或者通过redis的方式做中转。

**Comment by cloudwu** [comment]
一般是重要操作写本地文本日志文件。出现万一情况，现写个脚本也能提取出重要信息写回数据库，过程不需要自动化。

注意：是写本地文本日志文件。这样可以避免你用的数据库有 bug ，把数据弄坏掉了呢。毕竟只要是软件就有 bug ，说不定比停电还常见呢。

**Reply by ericzhanchina** [reply]
那就是任何一个协议过来，我们内存中修改之后，在日志中做记录，日志记录成功之后，然后返回给客户端协议结果，万一数据出了问题，再恢复本地文件的内容到数据库。写日志的速度会比写mysql的速度快吗？

**Reply by cloudwu** [reply]
只写你认为重要的，重要到万一停电了也要恢复的。而不是写所有的。

**Reply by ericzhanchina** [reply]
嗯，确实需要有一层保护
